src/main/scala/com/cloudera/sparkts/ARIMA.scala

```diff
@@ -0,0 +1,566 @@
+    val (p, d, q) = order
+    val addToLag = 1
+    val m = math.max(p, q) + addToLag // m > max(p, q)
+    // higher order AR(m) model
+    val arModel = Autoregression.fitModel(new DenseVector(diffedY), m)
+    val arTerms1 = Lag.lagMatTrimBoth(diffedY, m, includeOriginal = false)
+    val yTrunc = diffedY.drop(m)
+    val estimated = arTerms1.zip(
+      Array.fill(yTrunc.length)(arModel.coefficients)
+      ).map { case (v, b) => v.zip(b).map { case (yi, bi) => yi * bi}.sum + arModel.c }
+    // errors estimated from AR(m)
+    val errors = yTrunc.zip(estimated).map { case (y, yhat) => y - yhat }
+    // secondary regression, regresses X_t on AR and MA terms
+    val arTerms2 = Lag.lagMatTrimBoth(yTrunc, p, includeOriginal = false).drop(math.max(q - p, 0))
+    val errorTerms = Lag.lagMatTrimBoth(errors, q, includeOriginal = false).drop(math.max(p - q, 0))
+    val allTerms = arTerms2.zip(errorTerms).map { case (ar, ma) => ar ++ ma }
+    val regression = new OLSMultipleLinearRegression()
+    regression.setNoIntercept(!includeIntercept)
+    regression.newSampleData(yTrunc.drop(m - addToLag), allTerms)
+    val params = regression.estimateRegressionParameters()
+    params
+  }
+
+  /**
+   * Calculate a differenced vector of a given order. Size-preserving by leaving first `order`
+   * elements intact
+   * @param ts Series to difference
+   * @param order The difference order (e.g. x means y(i) = ts(i) - ts(i - x), etc)
+   * @return a new differenced vector
+   */
+  def differences(ts: Vector[Double], order: Int): Vector[Double] = {
+    if (order == 0) {
+      // for consistency, since we create a new vector in else-branch
+      ts.copy
+    } else {
+      val n = ts.length
+      val diffedTs = new DenseVector(Array.fill(n)(0.0))
+      var i = 0
+
+      while (i < n) {
+        // elements prior to `order` are copied over without modification
+        diffedTs(i) = if (i < order) ts(i) else ts(i) - ts(i - order)
+        i += 1
+      }
+      diffedTs
+    }
+  }
+
+  /**
+   * Calculate a "inverse-differenced" vector of a given order. Size-preserving by leaving first
+   * `order` elements intact
+   * @param ts Series to add up
+   * @param order The difference order to add (e.g. x means y(i) = ts(i) + y(i -
+   *              x), etc)
+   * @return a new vector where the difference operation as been inverted
+   */
+  def invDifferences(ts: Vector[Double], order: Int): Vector[Double] = {
+
+    if (order == 0) {
+      // for consistency, since we create a new vector in else-branch
+      ts.copy
+    } else {
+      val n = ts.length
+      val addedTs = new DenseVector(Array.fill(n)(0.0))
+      var i = 0
+
+      while (i < n) {
+        // elements prior to `order` are copied over without modification
+        addedTs(i) = if (i < order) ts(i) else ts(i) + addedTs(i - order)
+        i += 1
+      }
+      addedTs
+    }
+  }
+}
+
+class ARIMAModel(
+    val order:(Int, Int, Int), // AR(p), differencing(d), MA(q)
```

Comment:
I've broken it out in the `ARIMAModel` class, do you think the same should be done in the `ARIMA` object? I have kept it as a triple for now, since that is akin to how R's stats `arima` works, which takes a `c(p, d, q)` for order.